// @medusajs/workflows/dist/definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputAlias = exports.Workflows = void 0;
var Workflows;
(function (Workflows) {
    // Product workflows
    Workflows["CreateProducts"] = "create-products";
    Workflows["UpdateProducts"] = "update-products";
    // Cart workflows
    Workflows["CreateCart"] = "create-cart";
    Workflows["CreateInventoryItems"] = "create-inventory-items";
})(Workflows || (exports.Workflows = Workflows = {}));
var InputAlias;
(function (InputAlias) {
    InputAlias["Products"] = "products";
    InputAlias["ProductsInputData"] = "productsInputData";
    InputAlias["RemovedProducts"] = "removedProducts";
    InputAlias["InventoryItems"] = "inventoryItems";
    InputAlias["RemovedInventoryItems"] = "removedInventoryItems";
    InputAlias["AttachedInventoryItems"] = "attachedInventoryItems";
    InputAlias["DetachedInventoryItems"] = "detachedInventoryItems";
    InputAlias["InventoryItemsInputData"] = "inventoryItemsInputData";
})(InputAlias || (exports.InputAlias = InputAlias = {}));
