// @medusajs/link-modules/dist/utils/generate-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateGraphQLSchema = void 0;
const modules_sdk_1 = require("@medusajs/modules-sdk");
const utils_1 = require("@medusajs/utils");
const compose_link_name_1 = require("./compose-link-name");
function generateGraphQLSchema(joinerConfig, primary, foreign) {
    const fieldNames = primary.foreignKey.split(",").concat(foreign.foreignKey);
    const entityName = (0, utils_1.toPascalCase)("Link_" +
        (joinerConfig.databaseConfig?.tableName ??
            (0, compose_link_name_1.composeTableName)(primary.serviceName, primary.foreignKey, foreign.serviceName, foreign.foreignKey)));
    // Pivot table fields
    const fields = fieldNames.reduce((acc, curr) => {
        acc[curr] = {
            type: "String",
            nullable: false,
        };
        return acc;
    }, {});
    const extraFields = joinerConfig.databaseConfig?.extraFields ?? {};
    for (const column in extraFields) {
        fields[column] = {
            type: getGraphQLType(extraFields[column].type),
            nullable: !!extraFields[column].nullable,
        };
    }
    // Link table relationships
    const primaryField = `${(0, utils_1.camelToSnakeCase)(primary.alias)}: ${(0, utils_1.toPascalCase)((0, compose_link_name_1.composeTableName)(primary.serviceName))}`;
    const foreignField = `${(0, utils_1.camelToSnakeCase)(foreign.alias)}: ${(0, utils_1.toPascalCase)((0, compose_link_name_1.composeTableName)(foreign.serviceName))}`;
    let typeDef = `
    type ${entityName} {
      ${Object.entries(fields)
        .map(([field, { type, nullable }]) => `${field}: ${nullable ? type : `${type}!`}`)
        .join("\n      ")}
        
      ${primaryField}
      ${foreignField}
      
      createdAt: String!
      updatedAt: String!
      deletedAt: String
    }
  `;
    for (const extend of joinerConfig.extends ?? []) {
        const extendedModule = modules_sdk_1.MedusaModule.getModuleInstance(extend.serviceName);
        if (!extendedModule && !extend.relationship.isInternalService) {
            throw new Error(`Module ${extend.serviceName} not found. Please verify that the module is configured and installed, also the module must be loaded before the link modules.`);
        }
        const joinerConfig = modules_sdk_1.MedusaModule.getJoinerConfig(extend.serviceName);
        let extendedEntityName = joinerConfig?.linkableKeys?.[extend.relationship.primaryKey];
        if (!extendedEntityName) {
            continue;
        }
        extendedEntityName = (0, utils_1.toPascalCase)(extendedEntityName);
        const linkTableFieldName = (0, utils_1.camelToSnakeCase)((0, utils_1.lowerCaseFirst)(extend.relationship.alias));
        const type = extend.relationship.isList ? `[${entityName}]` : entityName;
        typeDef += `    
      extend type ${extendedEntityName} {
        ${linkTableFieldName}: ${type}
      }
    `;
    }
    return typeDef;
}
exports.generateGraphQLSchema = generateGraphQLSchema;
function getGraphQLType(type) {
    const typeDef = {
        numeric: "Float",
        integer: "Int",
        smallint: "Int",
        tinyint: "Int",
        mediumint: "Int",
        float: "Float",
        double: "Float",
        boolean: "Boolean",
        decimal: "Float",
        string: "String",
        uuid: "ID",
        text: "String",
        date: "Date",
        time: "Time",
        datetime: "DateTime",
        bigint: "BigInt",
        blob: "Blob",
        uint8array: "[Int]",
        array: "[String]",
        enumArray: "[String]",
        enum: "String",
        json: "JSON",
        jsonb: "JSON",
    };
    return typeDef[type] ?? "String";
}
